Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/EndRide.hs

```diff
@@ -540,7 +540,7 @@ recalculateFareForDistance ServiceHandle {..} booking ride recalcDistance' thres
               roundTrip = fromMaybe False booking.roundTrip,
               waitingTime = if isNothing ride.driverArrivalTime then Nothing else fmap (max 0) (secondsToMinutes . roundToIntegral <$> (diffUTCTime <$> ride.tripStartTime <*> (liftA2 max ride.driverArrivalTime (Just booking.startTime)))),
               stopWaitingTimes = stopsInfo <&> (\stopInfo -> max 0 (secondsToMinutes $ roundToIntegral (diffUTCTime (fromMaybe stopInfo.waitingTimeStart stopInfo.waitingTimeEnd) stopInfo.waitingTimeStart))),
-              actualRideDuration = roundToIntegral <$> (diffUTCTime <$> Just tripEndTime <*> ride.tripStartTime),
+              actualRideDuration = bool (roundToIntegral <$> (diffUTCTime <$> Just tripEndTime <*> ride.tripStartTime)) booking.estimatedDuration (passedThroughDrop && pickupDropOutsideOfThreshold && booking.tripCategory `elem` tripCategoryForNoRecalc && ride.distanceCalculationFailed == Just False), -- sending estimated duration in pass through cases as actual distance passed in such cases is estimated distance as well.
```

Comment:
`passedThroughDrop && pickupDropOutsideOfThreshold && booking.tripCategory `elem` tripCategoryForNoRecalc && ride.distanceCalculationFailed == Just False`
We can common this out and refactor as a seperate function as same logic is being repeated in 2 places. Let's create a common function that returns (distance, duration) based on the checks.